### PR TITLE
fix: prevent shares from being deposited to wrong address

### LINKED ISSUE #276: Shares can be indirectly transferred to 0x0
Users can bypass transfer protections to send shares to the 0x0 address burning them.

Users can deposit funds to obtain shares in the Yearn vaults. The shares are ERC20-compatible and can be transferred freely, except to the address of the contract or the 0x0 address:

### PR DESCRIPTION
**Pull request by steffenix**
fixes #276

### CODE REVIEW

File: contracts/Vault.vy [hunk 1]
```diff
@@ -846,6 +846,7 @@ def deposit(_amount: uint256 = MAX_UINT256, recipient: address = msg.sender) ->
     @return The issued Vault shares.
     """
     assert not self.emergencyShutdown  # Deposits are locked out
+    assert not (recipient in [self, ZERO_ADDRESS])
```
Comment (fubuloubu) [inline contracts/Vault.vy:849]: Instead, add to `erc20_safe_transfer` and `erc20_safe_transferFrom` instead

Comment (banteg) [reply contracts/Vault.vy:849]: `assert recipient not in [self, ZERO_ADDRESS]` would also work in latest Vyper

Comment (steffenix) [reply contracts/Vault.vy]: `erc20_safe_transferFrom` is used here and will fail:
https://github.com/yearn/yearn-vaults/blob/master/contracts/Vault.vy#L1718

Comment (steffenix) [reply contracts/Vault.vy]: @banteg https://github.com/yearn/yearn-vaults/pull/288

File: contracts/Vault.vy [hunk 2]
```diff
@@ -854,8 +855,7 @@ def deposit(_amount: uint256 = MAX_UINT256, recipient: address = msg.sender) ->
     if amount == MAX_UINT256:
         amount = min(
             self.depositLimit - self._totalAssets(),
-            self.token.balanceOf(msg.sender),
-        )
+            self.token.balanceOf(msg.sender))
```
Comment (fubuloubu) [inline contracts/Vault.vy:858]: keep this the way it was

File: contracts/Vault.vy [hunk 3]
```diff
@@ -855,7 +856,7 @@ def deposit(_amount: uint256 = MAX_UINT256, recipient: address = msg.sender) ->
         amount = min(
             self.depositLimit - self._totalAssets(),
             self.token.balanceOf(msg.sender),
-        )
+            )
```
Comment (fubuloubu) [inline contracts/Vault.vy:859]: Undo this change, `black` would make it look like this

**Review by fubuloubu — APPROVED** [review]
Seems like the `hardhat.config.js` file belongs in another PR?
